components/service/pocket/src/main/java/mozilla/components/service/pocket/spocs/SpocsUseCases.kt

```diff
@@ -0,0 +1,163 @@
+                return true
+            }
+
+            return false
+        }
+    }
+
+    /**
+     * Allows for querying the list of available Pocket sponsored stories.
+     *
+     * @param context [Context] used for various system interactions and libraries initializations.
+     * @param userId Unique identifier of the user which will be presented with sponsored stories.
+     * @param appId Unique identifier of the application using this feature.
+     */
+    internal inner class GetSponsoredStories(
+        private val context: Context,
+        private val userId: UUID,
+        private val appId: String,
+    ) {
+        /**
+         * Do an internet query for a list of Pocket sponsored stories.
+         */
+        suspend operator fun invoke(): List<PocketSponsoredStory> {
+            return getSpocsRepository(context).getAllSpocs()
+        }
+    }
+
+    /**
+     * Allows deleting all stored user data used for downloading sponsored stories.
+     *
+     * @param context [Context] used for various system interactions and libraries initializations.
+     * @param userId Unique user identifier previously used for downloading sponsored Pocket stories.
+     * @param appId Unique app identifier previously used for downloading sponsored Pocket stories.
+     */
+    internal inner class DeleteUserProfile(
+        private val context: Context,
+        private val userId: UUID,
+        private val appId: String,
+    ) {
+        /**
+         * Delete all stored user data used for downloading personalized sponsored stories.
+         */
+        suspend operator fun invoke(): Boolean {
+            val client = fetchClient
+            if (client == null) {
+                logger.error("Cannot delete user data. Service has incomplete setup")
+                return false
+            }
+
+            val provider = getSpocsProvider(client, userId, appId)
+            return when (provider.deleteProfile()) {
+                is Success -> {
+                    getSpocsRepository(context).deleteAllSpocs()
+                    true
+                }
+                is Failure -> {
+                    // Don't attempt to delete locally persisted stories to prevent mismatching issues
+                    // with profile deletion failing - applications still "showing it" but
+                    // with no sponsored articles to show.
+                    false
+                }
+            }
+        }
+    }
+
+    @VisibleForTesting
+    internal fun getSpocsRepository(context: Context) = SpocsRepository(context)
+
+    @VisibleForTesting
+    internal fun getSpocsProvider(client: Client, userId: UUID, appId: String) =
+        SpocsEndpoint.newInstance(client, userId, appId)
+
+    @VisibleForTesting
+    internal fun getPocketEndpoint(client: Client) = PocketEndpoint.newInstance(client)
+
+    internal companion object : NetworkUseCases by NetworkUseCase.Companion {
```

Comment:
See comment in above. Wouldn't it be easier if `(Pocket)NetworkUseCase` was an abstract class that's extended from by both use cases?

Reply:
`NetworkUseCase` is meant to be an abstract class by being open for extension with children inheriting the companion object. Actually marking it as `abstract` would indeed improve it - it would block instantiation of otherwise an usecase class without any functions to use.
The interface was there just to support delegation for the companion object and just because accessing the companion object of a parent is ugly.

Will look into avoiding the companion object altogether and use a custom factory to build the WorkerManager Worker classes instantiated with the appropriate usecases.

Reply:
As discussed in this thread - https://github.com/mozilla-mobile/android-components/pull/12070#pullrequestreview-959470575 it seems like the custom factory is a no-go and we'll still need to rely on some form of singletons for the WorkManager constructed workers to interact with any needed dependencies.
In which case the current approach of using static properties from an abstract class (marked properly as such) would be one possible implementation in that direction.